net/ip.c

```diff
@@ -0,0 +1,491 @@
+				  M0_NET_IP_AF_INET ? AF_INET : AF_INET6,
+				  &na->nip_ip_n.sn[0], ip_p, ARRAY_SIZE(ip_p));
+			rc = snprintf(buf, M0_NET_IP_STRLEN_MAX, "%s:%s:%s@%d",
+				      na->nip_fmt_pvt.ia.nia_family ==
+				      M0_NET_IP_AF_INET ? "inet" : "inet6",
+				      ip_protocol[na->nip_fmt_pvt.ia.nia_type],
+				      ip_p, na->nip_port);
+		} else
+			M0_LOG(M0_ERROR, "Format is currently not supported");
+	} else if (na->nip_format == M0_NET_IP_INET_HOSTNAME_FORMAT) {
+		inet_ntop(AF_INET, &na->nip_ip_n.sn[0], ip_p, ARRAY_SIZE(ip_p));
+		rc = m0_net_ip_to_hostname(ip_p, hostname);
+		if (rc != 0 )
+			return M0_ERR(rc);
+		rc = snprintf(buf, M0_NET_IP_STRLEN_MAX, "inet:%s:%s@%d", 
+			      ip_protocol[na->nip_fmt_pvt.ia.nia_type],
+			      hostname, na->nip_port);
+	}
+	if (rc <= 0 || rc >= M0_NET_IP_STRLEN_MAX)
+		return M0_ERR(-EINVAL);
+	M0_LOG(M0_DEBUG, "Address constructed: %s", buf);
+
+	return 0;
+}
+
+M0_INTERNAL int m0_net_hostname_to_ip(const char *hostname, char *ip,
+				      enum m0_net_ip_format *fmt)
+{
+	struct hostent  *hname;
+	struct in_addr **addr;
+	uint32_t         ip_n[4];
+	int              i;
+	int              n;
+	char            *cp;
+	char             name[M0_NET_IP_STRLEN_MAX] = {};
+
+	M0_ENTRY("Hostname=%s", (char*)hostname);
+	cp = strchr(hostname, '@');
+	if (cp == NULL)
+		return M0_ERR(-EINVAL);
+
+	n = cp - hostname;
+	name[n] = '\0';
+	strncat(name, hostname, n);
+
+	if (inet_pton(AF_INET, name, &ip_n[0]) == 1 ||
+	    inet_pton(AF_INET6, name, &ip_n[0]) == 1) {
+		/* Copy ip address as it is. */
+		*fmt = M0_NET_IP_INET_IP_FORMAT;
+		strcpy(ip, name);
+	} else {
+		*fmt = M0_NET_IP_INET_HOSTNAME_FORMAT;
+		hname = gethostbyname(name);
+		if (hname == NULL) {
+			M0_LOG(M0_ERROR, "gethostbyname err=%d for %s",
+			       h_errno, (char*)name);
+			/* Return positive rc for gethostbyname failure */
+			return M0_ERR(h_errno);
+		}
+		addr = (struct in_addr **)hname->h_addr_list;
+		for(i = 0; addr[i] != NULL; i++) {
+			/* Return the first one. */
+			strcpy(ip, inet_ntoa(*addr[i]));
+			M0_LOG(M0_DEBUG, "fqdn=%s ip=%s", (char*)name, ip);
+			return M0_RC(0);
+		}
+
+		/* If no valid addr structure found, then return error */
+		return M0_ERR(-errno);
+	}
+
+	return M0_RC(0);
+}
+
+M0_INTERNAL bool m0_net_ip_addr_eq(const struct m0_net_ip_addr *addr1,
+				   const struct m0_net_ip_addr *addr2, bool is_ncmp)
+{
+	M0_PRE(addr1 != NULL && addr2 != NULL);
+	if (!is_ncmp)
+		return (strcmp(addr1->nia_p, addr2->nia_p) == 0);
+	else
+		return (addr1->nia_n.nip_format == addr2->nia_n.nip_format &&
+			addr1->nia_n.nip_port == addr2->nia_n.nip_port     &&
+			/* For lnet address compare using m0_net_ip_la_eq(). */
+			((addr1->nia_n.nip_format == M0_NET_IP_LNET_FORMAT &&
+			m0_net_ip_la_eq(addr1, addr2))                    ||
+			/* For inet address compare using m0_net_ip_ia_eq(). */
+			(addr1->nia_n.nip_format != M0_NET_IP_LNET_FORMAT  &&
+			m0_net_ip_ia_eq(addr1, addr2))));
```

Comment:
This two lines of code 464 & 465 should be moved ahead of compare on line 461 & 462 will be optimization for Libfbaric.

((addr1->nia_n.nip_format == M0_NET_IP_LNET_FORMAT &&
  m0_net_ip_la_eq(addr1, addr2))

Reply:
It will have not much effect.
If `addr1->nia_n.nip_format == M0_NET_IP_LNET_FORMAT` from line 461 fails, then comparison on line 462 `m0_net_ip_la_eq(addr1, addr2)` will get ignored and condition on line 464 will get executed.